lib/utils/job-utils/HTTP-tool.js

```diff
@@ -0,0 +1,195 @@
+        this.settings = {};
+        
+        // ********** Helper Functions/Members **********
+        
+        /**
+        * Make sure that the method is a valid HTTP method
+        * @param {string} - testMethod: the intended method that for 
+        * http request.
+        * @return {boolean} - Whether the testmethod is valid
+        */
+        var validMethod = {'GET': true, 'PUT': true, 'POST':true, 
+            'DELETE':true, 'PATCH': true};
+        
+        this.isValidMethod = function(testMethod){
+            return (testMethod in validMethod);
+        };
+
+        this.isSettingValid = function(settings){
+            if (_.isEmpty(settings)) {
+                return false;
+            }
+            if ( ! ('url' in settings && 'method' in settings)) {
+                return false;
+            }
+            if (_.isEmpty(settings.url) || _.isEmpty(settings.method)) {
+                return false;
+            }
+            return true;
+        };
+    }
+
+    /**
+    * Set up the Request
+    * @param {object} settings - the object that provides the infomation 
+    * of all the works, example:
+    * {
+    *   url: "https://somewhat.website.com:some-port/some/path/to/file",
+    *   method: "GET"/"PUT"/"POST"/"DELETE"/"PATCH",
+    *   credential: {username: "foo", password: "bar"},
+    *   headers: {"token":"whatever-cookie-file"},
+    *   data: "Hello world",
+    *   verifySSL: false,
+    *   recvTimeoutMs: 2000
+    * }
+    * OR
+    * {
+    *   url: {
+    *       protocol: "https",
+    *       host: "somewhat.website.com",
+    *       port: "some-port",
+    *       path: "/some/path/to/file"
+    *   },
+    *   method: "GET"/"PUT"/"POST"/"DELETE"/"PATCH",
+    *   credential: {username: "foo", password: "bar"},
+    *   headers: {"token":"whatever-cookie-file"},
+    *   data: "Hello world",
+    *   verifySSL: false,
+    *   recvTimeoutMs: 2000
+    * }
+    *
+    * @return {Promise}
+    *   resolve() - do whatever next
+    *   reject(err) - do whatever you want with the err
+    */
+
+    HttpTool.prototype.setupRequest = function(settings) {
+        var self = this;
+        var err;
+        return new Promise(function(resolve, reject) {
+            if (! self.isSettingValid(settings)) {
+                err = new Error('Please provide at least url and'+
+                                     ' method to use HTTP tool!');
+                reject(err);
+            }
+            else {
+                if (self.isValidMethod(settings.method)) {
+                    self.settings = settings;
+                    resolve();
+                }
+                else {
+                    err = new Error('The method you provided is ' +
+                                        'not valid!');
+                    reject(err);
+                }
+            }});
+    };
+    
+    /**
+    * Run the request based on checked result
+    */ 
+    HttpTool.prototype.runRequest = function() {
```

Comment:
There is a lot of setup work for each request.  Could you do a one-time setup that is stored in the httpTool instance and break out the actual request method?